File: js_solutions/easy/climbing-stairs.js
// recursive solution
const fib = (n, memo = {}) => {
  if (n < 0) return 0;
  if (n === 0) return 1;

  memo[n] = memo[n] || fib(n - 1, memo) + fib(n - 2, memo);
  return memo[n];
}

const climbStairs = (num) => fib(num);

// iterative solution
const climbStairsIt = (num) => {
  let first = 1;
  let second = 1;
  while (num > 1) {
    [first, second] = [second, first + second];
    num--;
  }
  return second;
}

File: 01_session/easy/climbing_stairs.rb
# frozen_string_literal: true

# iterative solution
def climb_stairs(num)
  first = 1
  second = 1
  while num > 1
    temp = second
    second += first
    first = temp
    num -= 1
  end
  second
end

# recursive w/ memo
def fibonacci(num, memo = {})
  return 0 if n.negative?

  return 1 if n == 1

  memo[num] ||= fibonacci(num - 1, memo) + fibonacci(num - 2, memo)
end

def climb_stairs_rec(num)
  fibonacci(num + 1)
end
